demonstrations/tutorial_pulse_programming101.py

```diff
@@ -0,0 +1,383 @@
+
+Pulse programming in PennyLane
+------------------------------
+
+A user of a quantum computer typically operates on the higher and more abstract gate level.
+Future fault-tolerant quantum computers require this abstraction to allow for error correction.
+For noisy and intermediate sized quantum computers, the abstraction of decomposing quantum algorithms
+into a fixed native gate set can be a hindrance and unnecessarily increase execution time, therefore leading
+to more noise in the computation. The idea of differentiable pulse programming is to optimize quantum circuits on the pulse
+level with the aim of achieving the shortest interaction sequence a hardware system allows.
+
+In PennyLane, we can simulate arbitrary qubit system interactions to explore the possibilities of such pulse programs.
+First, we need to define the time-dependent Hamiltonian :math:`H(p, t)= \sum_i f_i(p, t) H_i` with constant operators :math:`H_i` and driving fields :math:`f_i(p, t)` that may
+depend on parameters :math:`p`. In PennyLane, we can do this in an intuitive way:
+"""
+
+import pennylane as qml
+import pennylane.numpy as np
+import jax.numpy as jnp
+import jax
+import matplotlib.pyplot as plt
+
+# remove jax CPU/GPU warning
+jax.config.update('jax_platform_name', 'cpu')
+
+def f1(p, t):
+    return jnp.polyval(p, t)
+def f2(p, t):
+    return p[0] * jnp.sin(p[1] * t)
+
+Ht = f1 * qml.PauliX(0) + f2 * qml.PauliY(1)
+
+##############################################################################
+# Note that when constructing such a Hamiltonian, the ``callable`` functions are 
+# expected to have the fixed signature ``(p, t)``, such that the Hamiltonian itself
+# can be called via ``H((p1, p2), t)``. 
+
+params = (jnp.ones(5), jnp.array([1., jnp.pi]))
+print(Ht(params, 0.5))
+
+##############################################################################
+# We can construct more complicated Hamiltonians like :math:`\sum_i X_i X_{i+1} + \sum_i f_i(p, t) Z_i` using :func:`qml.dot <pennylane.dot>`.
+# We use two sinusodials with random frequencies as the time-dependent parametrization for each :math:`Z_i`.
+
+coeffs = [jnp.array(1.)] * 2
+coeffs += [lambda p, t: jnp.sin(p[0]*t) + jnp.sin(p[1]*t) for _ in range(3)]
+ops = [qml.PauliX(i) @ qml.PauliX(i+1) for i in range(2)]
+ops += [qml.PauliZ(i) for i in range(3)]
+
+Ht = qml.dot(coeffs, ops)
+
+# random coefficients
+key = jax.random.PRNGKey(777)
+subkeys = jax.random.split(key, 3)
+params = [jax.random.uniform(subkeys[i], shape=[2], maxval=5) for i in range(3)]
+print(Ht(params, 0.5))
+
+##############################################################################
+# We can visualize the Hamiltonian interaction by plotting the time-dependent envelopes.
+
+ts = jnp.linspace(0., 5., 100)
+fs = Ht.coeffs_parametrized
+ops = Ht.ops_parametrized
+n_channels = len(fs)
+fig, axs = plt.subplots(nrows=n_channels, figsize=(5,2*n_channels))
```

Comment:
One _could_ consider removing the vertical space between the plots and the xticks of the upper two panels. Definitely a subjective preference question. I'll make suggestions for this, feel free to discard.
```suggestion
fig, axs = plt.subplots(nrows=n_channels, figsize=(5,2*n_channels), gridspec_kw={"hspace": 0})
```

Reply:
love it!